cluster-autoscaler/clusterstate/clusterstate.go

```diff
@@ -1090,3 +1115,114 @@ func fakeNode(instance cloudprovider.Instance) *apiv1.Node {
+}
+
+// CloudProviderNodeInstancesCache caches cloud provider node instances.
+type CloudProviderNodeInstancesCache struct {
+	sync.Mutex
+	cloudProviderNodeInstances map[string]*cloudProviderNodeInstancesCacheEntry
+	cloudProvider              cloudprovider.CloudProvider
+}
+
+// NewCloudProviderNodeInstancesCache creates new cache instance.
+func NewCloudProviderNodeInstancesCache(cloudProvider cloudprovider.CloudProvider) *CloudProviderNodeInstancesCache {
+	return &CloudProviderNodeInstancesCache{
+		cloudProviderNodeInstances: map[string]*cloudProviderNodeInstancesCacheEntry{},
+		cloudProvider:              cloudProvider,
+	}
+}
+
+func (cache *CloudProviderNodeInstancesCache) updateCacheEntryLocked(nodeGroup cloudprovider.NodeGroup, cacheEntry *cloudProviderNodeInstancesCacheEntry) {
+	cache.Lock()
+	defer cache.Unlock()
+	cache.cloudProviderNodeInstances[nodeGroup.Id()] = cacheEntry
+}
+
+func (cache *CloudProviderNodeInstancesCache) removeCacheEntryLocked(nodeGroup cloudprovider.NodeGroup) {
+	cache.Lock()
+	defer cache.Unlock()
+	delete(cache.cloudProviderNodeInstances, nodeGroup.Id())
+}
+
+func (cache *CloudProviderNodeInstancesCache) getCacheEntryLocked(nodeGroup cloudprovider.NodeGroup) (*cloudProviderNodeInstancesCacheEntry, bool) {
+	cache.Lock()
+	defer cache.Unlock()
+	cacheEntry, found := cache.cloudProviderNodeInstances[nodeGroup.Id()]
+	return cacheEntry, found
+}
+
+func (cache *CloudProviderNodeInstancesCache) removeEntriesForNonExistingNodeGroupsLocked(nodeGroups []cloudprovider.NodeGroup) {
+	cache.Lock()
+	defer cache.Unlock()
+	nodeGroupExists := map[string]bool{}
+	for _, nodeGroup := range nodeGroups {
+		nodeGroupExists[nodeGroup.Id()] = true
+	}
+	for nodeGroupId := range cache.cloudProviderNodeInstances {
+		if !nodeGroupExists[nodeGroupId] {
+			delete(cache.cloudProviderNodeInstances, nodeGroupId)
+		}
+	}
+}
+
+// GetCloudProviderNodeInstances returns cloud provider node instances for all node groups returned by cloud provider.
+func (cache *CloudProviderNodeInstancesCache) GetCloudProviderNodeInstances() (map[string][]cloudprovider.Instance, error) {
+	results := map[string][]cloudprovider.Instance{}
+	for _, nodeGroup := range cache.cloudProvider.NodeGroups() {
+		nodeGroupInstances, err := cache.GetCloudProviderNodeInstancesForNodeGroup(nodeGroup)
+		if err != nil {
+			return nil, err
+		}
+		results[nodeGroup.Id()] = nodeGroupInstances
+	}
+	return results, nil
+}
+
+// GetCloudProviderNodeInstancesForNodeGroup returns cloud provider node instances for the given node group.
+func (cache *CloudProviderNodeInstancesCache) GetCloudProviderNodeInstancesForNodeGroup(nodeGroup cloudprovider.NodeGroup) ([]cloudprovider.Instance, error) {
+	cacheEntry, found := cache.getCacheEntryLocked(nodeGroup)
+	if found {
+		if cacheEntry.isStale() {
+			klog.Warningf("Entry cloudProviderNodeInstances is stale, refresh time is %v", cacheEntry.refreshTime)
```

Comment:
If we are doing refresh every two minutes and freshness interval is 3 minutes we will be seeing a lot of those.  If we dozens/hundreds of node groups we will not manage to refresh all within 1 minut (3-1)

Reply:
changed